### PR TITLE
Integrate ticket auto buyer

### PR DESCRIPTION
**Pull request by vctt94**
This PR depends on https://github.com/decred/dcrwallet/pull/1786

it starts integrating the ticket autobuyer into decrediton using the new vsp.

It still a WIP because right now because we need to hard code responses in order to get new VSPs. This can be fixed after https://github.com/decred/dcrwebapi/pull/104 is merged.

### CODE REVIEW

File: app/actions/VSPActions.js [hunk 1]
```diff
@@ -23,6 +23,41 @@ export const getVSPInfo = (host) => async (dispatch) => {
...
+  dispatch({ type: GETVSPTICKETSTATUS_ATTEMPT });
+  try {
+    const timestamp = Math.trunc(new Date()/1000);
+    const request = JSON.stringify({
+      timestamp: timestamp,
+      tickethash: tickethash
+    });
+    const signature = await dispatch(getTicketSignature(tickethash, request, passphrase));
+    // host here needs "http://" or "https://". When sending it to dcrwallet it can not have.
+    const host = "https://teststakepool.decred.org"
```
Comment (amassarwi) [inline app/actions/VSPActions.js:41]: lint

File: app/components/inputs/LEGACY_StakePoolSelect.js [hunk 2]
```diff
@@ -1,7 +1,6 @@
 import { Creatable } from "react-select";
 import Select from "react-select";
 import { FormattedMessage as T, injectIntl, defineMessages } from "react-intl";
-import { newStakePool } from "connectors";
```
Comment (amassarwi) [inline app/components/inputs/LEGACY_StakePoolSelect.js:4]: 🔪

File: app/components/inputs/VSPSelect.js [hunk 3]
```diff
@@ -3,10 +3,10 @@ import { injectIntl, defineMessages } from "react-intl";
 import { useState } from "react";
 import * as vspa from "actions/VSPActions";
 import { useDispatch } from "react-redux";
+import { base64ToHex } from "helpers";
```
Comment (amassarwi) [inline app/components/inputs/VSPSelect.js:6]: lint

File: app/components/views/TicketsPage/PurchaseTab/LEGACY_PurchasePage/LEGACY_TicketAutoBuyer/index.js [hunk 4]
```diff
@@ -1,4 +1,3 @@
-import ticketAutoBuyer from "connectors/ticketAutoBuyer";
```
Comment (amassarwi) [inline app/components/views/TicketsPage/PurchaseTab/LEGACY_PurchasePage/LEGACY_TicketAutoBuyer/index.js:1]: 🔪

File: app/components/views/TicketsPage/PurchaseTab/PurchaseTickets/PurchaseTickets.jsx [hunk 5]
```diff
@@ -9,19 +12,23 @@ const Tickets = ({ toggleIsLegacy }) => {
     sidebarOnBottom,
     isWatchingOnly,
     discoverAvailableVSPs,
+    availableVSPs,
     defaultSpendingAccount,
+    // spendingAccounts,
```
Comment (amassarwi) [inline app/components/views/TicketsPage/PurchaseTab/PurchaseTickets/PurchaseTickets.jsx:17]: comment

File: app/components/views/TicketsPage/PurchaseTab/PurchaseTickets/PurchaseTickets.jsx [hunk 6]
```diff
@@ -9,19 +12,23 @@ const Tickets = ({ toggleIsLegacy }) => {
...
     ticketPrice
   } = usePurchaseTab();
+  useMountEffect(discoverAvailableVSPs);
```
Comment (amassarwi) [inline app/components/views/TicketsPage/PurchaseTab/PurchaseTickets/PurchaseTickets.jsx:20]: 😍

File: app/components/views/TicketsPage/PurchaseTab/PurchaseTickets/PurchaseTickets.jsx [hunk 7]
```diff
@@ -9,19 +12,23 @@ const Tickets = ({ toggleIsLegacy }) => {
...
 
   const [account, setAccount] = useState(defaultSpendingAccount);
+  // const [vsp, setVSP] = useState(null);
```
Comment (amassarwi) [inline app/components/views/TicketsPage/PurchaseTab/PurchaseTickets/PurchaseTickets.jsx:23]: comment

File: app/components/views/TicketsPage/PurchaseTab/TicketAutoBuyer/Form.jsx [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+            onSubmit={onStartAutoBuyer}
+          />
+        )}
+        <div className="stakepool-auto-buyer-row-portion-half">
+          <div className="stakepool-autobuyer-label">
+            <T id="autobuyer.accountFrom" m="From" />:
+          </div>
+          <div className="stakepool-autobuyer-input">
+            <AccountsSelect
+              {...{ account }}
```
Comment (amassarwi) [inline app/components/views/TicketsPage/PurchaseTab/TicketAutoBuyer/Form.jsx:82]:
```suggestion
              account={account}
```

File: app/components/views/TicketsPage/PurchaseTab/TicketAutoBuyer/TicketAutoBuyer.jsx [hunk 9]
```diff
@@ -0,0 +1,60 @@
+import { substruct, compose, eq, get } from "fp";
```
Comment (amassarwi) [inline app/components/views/TicketsPage/PurchaseTab/TicketAutoBuyer/TicketAutoBuyer.jsx:1]: linting

File: app/components/views/TicketsPage/PurchaseTab/TicketAutoBuyer/TicketAutoBuyer.jsx [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    setBalanceToMaintain(atomValue)
+  }
+
+  const onStartAutoBuyer = (passphrase) => {
+    onEnableTicketAutoBuyer(passphrase, account, balanceToMaintain, vsp);
+  }
+
+  const onStopAutoBuyer = () => onDisableTicketAutoBuyer();
+
+  // TODO remove this - using to test
```
Comment (amassarwi) [inline app/components/views/TicketsPage/PurchaseTab/TicketAutoBuyer/TicketAutoBuyer.jsx:33]: reminder ;)

File: app/middleware/vspapi.js [hunk 11]
```diff
@@ -22,18 +31,74 @@ const POST = (path, apiToken, json) => {
...
+  // This json request is strigfied at the call which is making it.
+  return axios.post(path, json, config);
+};
+
 // stakepPoolInfoResponseToConfig converts a response object for the
 // stakePoolInfo call into an object array of available stakepool configs.
 function stakepPoolInfoResponseToConfig(response) {
-  const stakePoolNames = Object.keys(response.data);
+    // TODO remove this json.
+    // used to test with localhost and teststakepool
```
Comment (amassarwi) [inline app/middleware/vspapi.js:48]: reminder to remove

File: app/middleware/vspapi.js [hunk 12]
```diff
@@ -22,18 +31,74 @@ const POST = (path, apiToken, json) => {
...
   return stakePoolNames
     .map((name) => {
-      const { APIEnabled, URL, Network, APIVersionsSupported } = response.data[
+      const { APIEnabled, URL, Network, APIVersionsSupported } = data[
         name
       ];
       return !APIEnabled ? null : { Host: URL, Network, APIVersionsSupported };
     })
     .filter((v) => v);
+  // const stakePoolNames = Object.keys(response.data);
```
Comment (amassarwi) [inline app/middleware/vspapi.js:93]: comments

**Review by alexlyp — APPROVED** [review]

**Pull request merged**